service/controller.go

```diff
@@ -1513,10 +1513,122 @@ func (s *service) ListVolumes(_ context.Context,
+			Snapshot: s.getCSISnapshot(snap.Name, snap.Path, snap.Created, snap.Size),
+		}
+	}
+
+	log.Debugf("Returning snapshot list: count=%d, nextToken=%s", len(entries), nextToken)
+	return &csi.ListSnapshotsResponse{
+		Entries:   entries,
+		NextToken: nextToken,
+	}, nil
+}
+
+func (s *service) listPowerScaleSnapshots(ctx context.Context, startToken, maxEntries int, snapID, srcID string) (isi.SnapshotList, string, error) {
+	ctx, log, _ := GetRunIDLog(ctx)
+	log.Infof("Entering listPowerScaleSnapshots: snapID=%s, srcID=%s", snapID, srcID)
+
+	var allSnapshots isi.SnapshotList
+
+	for _, config := range s.getIsilonClusters() {
+		log := log.WithField("cluster", config.ClusterName)
+		snapshots, err := config.isiSvc.GetSnapshots(ctx)
+		if err != nil {
+			log.Errorf("Failed to get snapshots: %v", err)
+			continue
+		}
+		log.Debugf("Fetched snapshots: count=%d", len(snapshots))
+
+		for _, snap := range snapshots {
+			if shouldIncludeSnapshot(ctx, snap, snapID, srcID) {
+				log.Debugf("Including snapshot: ID=%d", snap.ID)
+				normalizeSnapshot(ctx, snap, config, s)
+				allSnapshots = append(allSnapshots, snap)
+			}
+		}
+	}
+
+	if startToken > len(allSnapshots) {
+		err := fmt.Errorf("startingToken=%d > len(allSnapshots)=%d", startToken, len(allSnapshots))
+		log.Errorf("Invalid starting token: %v", err)
+		return nil, "", status.Errorf(codes.Aborted, "invalid starting token, error: %s", err.Error())
+	}
+
+	remaining := len(allSnapshots) - startToken
+	if maxEntries == 0 || maxEntries > remaining {
+		maxEntries = remaining
+	}
+
+	nextToken := ""
+	if startToken+maxEntries < len(allSnapshots) {
+		nextToken = fmt.Sprintf("%d", startToken+maxEntries)
+	}
+
+	log.Infof("Returning snapshot slice: start=%d, count=%d, nextToken=%s", startToken, maxEntries, nextToken)
+	return allSnapshots[startToken : startToken+maxEntries], nextToken, nil
+}
+
+func shouldIncludeSnapshot(ctx context.Context, snap isi.Snapshot, snapID, srcID string) bool {
+	if snapID != "" {
+		id, _, _, _ := id.ParseNormalizedSnapshotID(ctx, snapID)
+		return strconv.FormatInt(snap.ID, 10) == id
+	}
+	if srcID != "" {
+		srcName, _, _, _, _ := id.ParseNormalizedVolumeID(ctx, srcID)
+		return strings.EqualFold(srcName, isilonfs.GetVolumeNameFromExportPath(snap.Path))
+	}
+	return true
+}
+
+func normalizeSnapshot(ctx context.Context, snap isi.Snapshot, config *IsilonClusterConfig, s *service) {
+	ctx, log, _ := GetRunIDLog(ctx)
+	log = log.WithField("snapshotID", snap.ID)
+	log.Debugf("Normalizing snapshot: ID=%d", snap.ID)
+
+	snap.Name = id.GetNormalizedSnapshotID(ctx, strconv.FormatInt(snap.ID, 10), config.ClusterName, s.opts.AccessZone)
+	volName := isilonfs.GetVolumeNameFromExportPath(snap.Path)
+	export, _ := config.isiSvc.GetExportWithPath(ctx, snap.Path)
+
+	if export != nil {
+		snap.Path = id.GetNormalizedVolumeID(ctx, volName, export.ID, export.Zone, config.ClusterName)
+		log.Debugf("Normalized with export: exportID=%d, zone=%s", export.ID, export.Zone)
+	}
```

Comment:
won't there be any errors here? What happens if there is any error for ex: if the export is nil, do we consider those objects as is?